src/material-experimental/mdc-list/testing/list-harness-base.ts

```diff
@@ -0,0 +1,153 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {
+  ComponentHarness,
+  ComponentHarnessConstructor,
+  HarnessPredicate
+} from '@angular/cdk/testing';
+import {DividerHarnessFilters, MatDividerHarness} from '@angular/material/divider/testing';
+import {BaseListItemHarnessFilters, SubheaderHarnessFilters} from './list-harness-filters';
+import {MatSubheaderHarness} from './list-item-harness-base';
+
+/** Represents a section of a list falling under a specific header. */
+export interface ListSection<I> {
+  /** The heading for this list section. `undefined` if there is no heading. */
+  heading?: string;
+
+  /** The items in this list section. */
+  items: I[];
+}
+
+/**
+ * Shared behavior among the harnesses for the various `MatList` flavors.
+ * @template T A constructor type for a list item harness type used by this list harness.
+ * @template C The list item harness type that `T` constructs.
+ * @template F The filter type used filter list item harness of type `C`.
+ * @docs-private
+ */
+export class MatListHarnessBase<
+    T extends(ComponentHarnessConstructor<C>& {with: (options?: F) => HarnessPredicate<C>}),
+             C extends ComponentHarness, F extends BaseListItemHarnessFilters> extends
+    ComponentHarness {
+  protected _itemHarness: T;
+
+  /**
+   * Gets a list of harnesses representing the items in this list.
+   * @param filters Optional filters used to narrow which harnesses are included
+   * @return The list of items matching the given filters.
+   */
+  async getItems(filters?: F): Promise<C[]> {
+    return this.locatorForAll(this._itemHarness.with(filters))();
+  }
+
+  /**
+   * Gets a list of `ListSection` representing the list items grouped by subheaders. If the list has
+   * no subheaders it is represented as a single `ListSection` with an undefined `heading` property.
+   * @param filters Optional filters used to narrow which list item harnesses are included
+   * @return The list of items matching the given filters, grouped into sections by subheader.
+   */
+  async getItemsGroupedBySubheader(filters?: F): Promise<ListSection<C>[]> {
+    const listSections = [];
+    let currentSection: ListSection<C> = {items: []};
+    const itemsAndSubheaders =
+        await this.getItemsWithSubheadersAndDividers({item: filters, divider: false});
+    for (const itemOrSubheader of itemsAndSubheaders) {
```

Comment:
should this `for` be refactored into a `Promise.all`?

Reply:
Makes sense. I've updated the non-MDC list harness to also perform the header retrieval concurrently.